[BEAM-4798] Fix IndexOutOfBoundsException in Flink runner

aromanenko-dev opened the pull request:
Fixes the issue when number of IO readers is less than value of Flink parallelism (number of task slots). In this case, extra task slots will be idling. 

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

aromanenko-dev commented:
@aljoscha Could you take a look on this, please?

aljoscha commented:
Run Flink ValidatesRunner

aljoscha commented on runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapper.java:
```diff
@@ -202,7 +202,15 @@ public void run(SourceContext<WindowedValue<ValueWithRecordId<OutputT>>> ctx) th
+      // parallelism is 2 and number of Kafka topic partitions is 1). In this case, we just idle
+      // this executor.
+      LOG.warn("Number of readers is 0, idle");
+      while (isRunning) {
```
we can just let this fall through to the idle loop at the end

aromanenko-dev replied:
Agree, fixed

aljoscha commented on runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapper.java:
```diff
@@ -202,7 +202,15 @@ public void run(SourceContext<WindowedValue<ValueWithRecordId<OutputT>>> ctx) th
+      // It can happen when value of parallelism is greater than number of IO readers (for example,
+      // parallelism is 2 and number of Kafka topic partitions is 1). In this case, we just idle
+      // this executor.
+      LOG.warn("Number of readers is 0, idle");
```
Probably shouldn't even be a warning, since it's a completely normal thing that can just happen

aromanenko-dev commented:
@aljoscha Thank you for review. I addressed your suggestions.
Regarding a unit test. I agree that it would make sense to add a test for this case but it’s not clear for me how to properly do that. Actually, as I see in `UnboundedSourceWrapperTest.UnboundedSourceWrapperTestWithParams.data()` it creates an array with different combinations of number of tasks and splits. And there are already combinations where `numTasks > numSplits`, which is a case of this PR issue.
So, I suppose, it’s already should be tested but it didn’t fail before? Or I missed something?

aaltay commented:
@aljoscha could you take another look at this please?

The current version seems to be only adding a new log statement, and it does not look like it will have unintended consequences.

aljoscha commented:
@aromanenko-dev So is the exception you mention in https://issues.apache.org/jira/browse/BEAM-4798 not an issue anymore?

aromanenko-dev commented:
@aljoscha Yes, this PR fixes this issue. I checked it on my local Flink cluster with `parallelism=2` and different number of Kafka partitions - 1, 2 and 3

aaltay commented:
Adding a unit test makes sense to me. 

@aromanenko-dev are you unblocked? Do you need help?

aromanenko-dev commented:
@aljoscha @aaltay Sorry for delay with response, I was on vacation for last two weeks. Now I'm back and, in the first order, I'd like to find out why already existed tests were not failed because they are supposed to cover this test case (when the number of IO readers is less than value of Flink parallelism value).

aljoscha commented:
@aromanenko-dev No problem at all. 😄 

I wrote this already to @aromanenko-dev via Slack but the problem seems to be that `UnboundedSourceWrapperTestWithParams` doesn't match the test inclusion pattern so it's never executed.

aromanenko-dev commented:
@aljoscha Thank you! As we discussed on Slack, after renaming the test class name, I've managed to run parameterized tests but it they still pass without my fix. Although, some of them have to fail where `numSplits < numTasks`
